tensorbay/label/tests/test_subcatalog.py

```diff
@@ -0,0 +1,213 @@
+
+
+class TestCommnoSubcatalog:
+    """This calss used to test
+    Box2DSubcatalog,
+    Box3DSubcatalog,
+    Polygon2DSubcatalog,
+    Polyline2DSubcatalog
+    """
+
+    @pytest.mark.parametrize(
+        "COMMON_SUBCATALOG",
+        [Box2DSubcatalog, Box3DSubcatalog, Polygon2DSubcatalog, Polyline2DSubcatalog],
+    )
+    def test_init_subclass(self, COMMON_SUBCATALOG):
+        subcatalog = COMMON_SUBCATALOG()
+        assert subcatalog._supports == (
+            SupportIsTracking,
+            SupportCategories,
+            SupportAttributes,
+        )
+
+    @pytest.mark.parametrize(
+        "COMMON_SUBCATALOG",
+        [Box2DSubcatalog, Box3DSubcatalog, Polygon2DSubcatalog, Polyline2DSubcatalog],
+    )
+    def test_loads(self, COMMON_SUBCATALOG):
+        box2d_subcatalog = COMMON_SUBCATALOG.loads(_COMMON_SUBCATALOG_CONTENT)
+
+        assert box2d_subcatalog.is_tracking == True
+        assert list(box2d_subcatalog.categories) == ["0", "1"]
+        assert box2d_subcatalog.attributes["gender"].enum == ["male", "female"]
+
+    @pytest.mark.parametrize(
+        "COMMON_SUBCATALOG",
+        [Box2DSubcatalog, Box3DSubcatalog, Polygon2DSubcatalog, Polyline2DSubcatalog],
+    )
+    def test_dumps(self, COMMON_SUBCATALOG, categories, attributes):
+        subcatalog = COMMON_SUBCATALOG()
+        subcatalog.is_tracking = True
+        subcatalog.categories = categories
+        subcatalog.attributes = attributes
+        assert subcatalog.dumps() == _COMMON_SUBCATALOG_CONTENT
+
+
+class TestKeypoints2DSubcatalog:
+    def test_init_subclass(self):
+        keypoints2d_subcatalog = Keypoints2DSubcatalog()
+        assert keypoints2d_subcatalog._supports == (
+            SupportIsTracking,
+            SupportCategories,
+            SupportAttributes,
+        )
+
+    def test_loads(self):
+        keypoints2d_subcatalog = Keypoints2DSubcatalog.loads(_KEYPOINTS2D_SUBCATALOG)
+
+        assert keypoints2d_subcatalog.is_tracking == True
+        assert list(keypoints2d_subcatalog.categories) == ["0", "1"]
+        assert keypoints2d_subcatalog.attributes["gender"].enum == ["male", "female"]
+
+    def test_dumps(self, categories, attributes, keypoints):
+        keypoints2d_subcatalog = Keypoints2DSubcatalog()
+        keypoints2d_subcatalog.is_tracking = True
+        keypoints2d_subcatalog.categories = categories
+        keypoints2d_subcatalog.attributes = attributes
+        keypoints2d_subcatalog._keypoints = [keypoints]
+        assert keypoints2d_subcatalog.dumps() == _KEYPOINTS2D_SUBCATALOG
+
+    def test_add_keypoints(self, keypoints):
+        keypoints2d_subcatalog = Keypoints2DSubcatalog()
+        keypoints2d_subcatalog.add_keypoints(
+            5,
+            names=[
+                "L_shoulder",
+                "L_Elbow",
+                "L_wrist",
+                "R_Shoulder",
+                "R_Elbow",
+            ],
+            skeleton=[(0, 1), (1, 2), (3, 4), (4, 5)],
+        )
```

Comment:
This keypoints code block seems appears many times, is it possible to refine it?